stellar-contract-env-host/src/host.rs

```diff
@@ -299,15 +299,31 @@ impl Env for Host {
     }
 
     fn obj_to_u64(&mut self, u: RawVal) -> u64 {
-        todo!()
+        unsafe {
+            self.unchecked_visit_val_obj(u, move |ho| {
+                if let Some(HostObject::U64(num)) = ho {
```

Comment:
I see your point. If I understood correct, the assumption host is making is the `u: RawVal` is a `RawObj`, by calling the unsafe method `unchecked_visit_val_obj`. If `u` is not `Tag::Object`, the result from this method will match to `None` which triggers the the `!panic`. So it feels to me ok, unless my understanding is missing. 
My first implementation (in the 1st commit which you also reviewed), goes from `RawVal -> RawObj -> ScObject->u64`, which doesn't involve the unsafe method but takes an extra hop. I then saw a couple of host functions already exist (`vec_len` and `vec_push`) that uses the `unchecked_visit_val_obj` so I decided to change my implementation. Do you think that was a safer approach?

Reply:
I think a couple things here:

1. Host should probably have a systematic treatment for "fetch an object by ID with an expected host object type, and fail if it's the wrong type" so we don't have to repeat that pattern everywhere.
2. In general it should not be possible for guest code to get the host to panic -- this would make it possible for a smart contract to crash stellar-core -- so we need a general pattern where the host function implementation is actually a partial function returning Result<...> and the entrypoint from Env panics on the error case when running locally / linking directly to the SDK and traps the VM when being invoked through a call from the VM. This is a little subtle and may involve some changes to the trait and method signatures. I'll try to put something together.

Reply:
> 2\. This is a little subtle and may involve some changes to the trait and method signatures. I'll try to put something together.

I just merged https://github.com/stellar/rs-stellar-contract-env/pull/39 that changed the signatures, so recommend updating your local if you just started doing this!